daemonize.1
.\" -*-nroff-*-
.\"
.\" With the exception of the "install-sh" script and the "getopt.c" source,
.\" this software is released under a BSD-style license. See the LICENSE
.\" file for details.
.\"
.\" Copyright (c) 2003-2006 Brian M. Clapper, bmc <at> clapper <dot> org
.\"
.\" $Id$
.\" ---------------------------------------------------------------------------
.\"
.ie n .ds Bu o
.el   .ds Bu \(bu
.\"
.TH daemonize 1 "August 2006" Unix "User Manuals"
.SH NAME
.PP
daemonize \- run a program as a Unix daemon
.SH SYNOPSIS
.PP
.B daemonize
[-a]
[-c \fIdirectory\fP]
[-e \fIstderr\fP]
[-o \fIstdout\fP]
[-p \fIpidfile\fP] 
[-l \fIlockfile\fP]
[-u \fIuser\fP]
[-v]
.I "path [arg] ..."
.SH DESCRIPTION
.PP
.B daemonize
runs a command as a Unix daemon. 
As defined in W. Richard Stevens' 1990 book,
.I "Unix Network Programming"
(Addison-Wesley, 1990), a daemon is
\*(lqa process that executes `in the background' (i.e., without an associated
terminal or login shell) either waiting for some event to occur, or waiting
to perform some specified task on a periodic basis.\*(rq Upon startup, a
typical daemon program will:
.IP \*(Bu 2
Close all open file descriptors (especially standard input, standard
output and standard error)
.IP \*(Bu 2
Change its working directory to the root filesystem, to ensure that
it doesn't tie up another filesystem and prevent it from being unmounted
.IP \*(Bu 2
Reset its
.I umask
value
.IP \*(Bu 2
Run in the background (i.e.,
.IR fork )
.IP \*(Bu 2
Disassociate from its process group (usually a shell), to insulate
itself from signals (such as HUP) sent to the process group
.IP \*(Bu 2
Ignore all terminal I/O signals
.IP \*(Bu 2
Disassociate from the control terminal (and take steps not to reacquire one)
.IP \*(Bu 2
Handle any
.SM SIGCLD
signals
.PP
Most programs that are designed to be run as daemons do that work for
themselves. However, you'll occasionally run across one that does not.
When you must run a daemon program that does not properly make
itself into a true Unix daemon, you can use 
.B daemonize
to force it to run as a true daemon.
.SH OPTIONS
.PP
.IP -a
Append to the output files, rather than overwriting them (which is the
default). Only applicable if
.I -e
and/or
.I -o
are specified.
.\"
.IP "-c \fIdirectory\fP"
Specifies the directory to which to change before running the program.
Defaults to "/". The choice for this option is important. The file system
containing the daemon's working directory cannot be unmounted while the
daemon is running. That's why most daemons are careful to use a working
directory on the root file system.
.\"
.IP "-e \fIstderr\fP"
Redirect standard error to the specified file, instead of "/dev/null".
.sp
.B Warning:
Be careful where you redirect the output! The file system containing the
open file cannot be unmounted as long as the file is open. For best
results, make sure that this output file is on the same file system as
the daemon's working directory. (See the
.I -c
option.)
.\"
.IP "-E \fIname=value\fP"
Add an environment variable to the environment the daemon will see. The
parameter must be of the form
.I name=value.
This parameter may be specified multiple times.
.\"
.IP "-o \fIstdout\fP"
Redirect standard output to the specified file, instead of "/dev/null".
.sp
.B Warning:
Be careful where you redirect the output! The file system containing the
open file cannot be unmounted as long as the file is open. For best
results, make sure that this output file is on the same file system as
the daemon's working directory. (See the
.I -c
option.)
.\"
.IP "-p \fIpidfile\fP"
Causes 
.B daemonize
to write the numeric process ID (PID) of the running daemon to the
specified file. This option is useful when the program being daemonized
doesn't create its own PID file.
.IP "-l \fIlockfile\fP"
Single-instance checking. Causes
.B daemonize
to ensure that no more than one instance of the daemon is running by
placing an exclusive lock on given lockfile. If another process already has
a lock on the lockfile,
.I daemonize
exits.
.PP
It 
.I is 
possible to use the
.I pidfile
as the lock file (e.g., "-p /var/run/foo -l /var/run/foo"),
though typical daemons use separate files.
.PP
NOTE: If the executed program decides to close all file descriptors, 
the single-instance locking will not work, since the lock depends on an
open file descriptor. (The operating system kernel removes the lock once
the process holding the lock closes the file or exits.)
Normal processes that do not daemonize themselves do not
usually close all file descriptors.
.\"
.IP "-u \fIuser\fP"
Run the program as the specified user. This option only works if
.B daemonize
is invoked by the superuser.
Note: For obvious reasons, it's very dangerous to install
.B daemonize
as a
.I setuid-to-root
executable. For that reason, 
.B daemonize
will refuse to run if it detects that it has been
installed that way.
.\"
.IP -v
Cause
.B daemonize
to write verbose messages to standard error, telling what it's doing
as it daemonizes the program.
.SH NOTES
.PP
If the host operating system provides the
.BR daemon (3)
library routine, 
.B daemonize
will use it. Otherwise,
.B daemonize
uses its own version of
.BR daemon (3).
This choice is made at compile time. (BSD 4.4-derived operating systems
tend to provide their own
.BR daemon (3)
routine.)
.PP
FreeBSD 5.0 introduced a
.BR daemon (1)
command that is similar to, but less functional, than
.BR daemonize .
.SH LICENSE
.PP
This program released under a BSD-style license. For more details,
consult the LICENSE file accompanying the source distribution, or
visit "http://software.clapper.org/daemonize/LICENSE".
.SH SEE ALSO
.PP
.BR daemon (3),
.BR setsid (2),
.BR flock(2)
.PP
.B daemonize
Home Page: http://software.clapper.org/daemonize/
.SH AUTHOR
.PP
Brian M. Clapper,
.I "bmc <at> clapper <dot> org"
.SH CONTRIBUTORS
.PP
Support for the
.I -e
and
.I -o
options is based on a patch submitted by
Tim Starling
.I "(tstarling <at> wikimedia <dot> org)."
.PP
Support for the
.I -l
option is based on a patch submitted by
Yakov Lerner
.I "(iler.ml <at> gmail <dot> com)."


CHANGELOG
daemonize Change Log
====================

Changes in version 1.6.1:

- Makefile now generates daemonize.html, an HTML-ized version of the man page.

-----------------------------------------------------------------------------
Changes in version 1.6:

Rolled in changes submitted by Ken D. Farnen, (kenf /at/ 14Cubed.COM):

- On slackware, all of the rc.???? scripts run directly from init, and init
  doesn't set up much in the way of an environment, whereas login stuffs
  loads of stuff into the environment. Daemonize now ensures that a minimum
  environment is passed to the exec-ed daemon (IFS and PATH).

- The -u flag now puts USER, HOME and LOGNAME into the environment, for the
  benefit of programs that rely on them. The lack of these environment
  variables seems to have deleterious effects on MythTV, for example.

- On Slackware Linux, access to video and audio device entries is
  controlled by secondary group membership. (e.g., A user who needs access
  to a dvb device needs to be in group "video", and a user who needs access
  to dsp devices needs to be in group "audio".), This is typically done by
  the login code. Daemonize now initializes those group memberships, as
  well, if the initgroups() function is present.

-----------------------------------------------------------------------------
Changes in version 1.5.6:

- Accepted and applied a patch from Gary T. Giesen (giesen /at/ snickers.org)
  that allows the daemonize Makefile to play well with Linux RPMs.

-----------------------------------------------------------------------------
Changes in version 1.5.5:

- Changes in 1.5.4 made it impossible to create PID files in directories
  not owned by the user specified with -u. Prior to 1.5.4, doing so was
  possible. It is now possible again, and the functionality has been
  corrected slightly to ensure that the proper PID is written and that the
  PID file has mode 0644.

-----------------------------------------------------------------------------
Changes in version 1.5.4:

- If a user name is specified via the -u option, the output files specified
  by the -o and -e options are now opened *after* daemonize switches users,
  not before (as happened in previous versions). This change ensures that
  the output files end up being owned by that user, not by root. This
  change has no effect if -u is not used.

- Rearranged and simplified some of the code in daemonize.c

- Added some documentation to the code.

-----------------------------------------------------------------------------
Changes in version 1.5.3:

- Minor licensing change. (configure.in still emitted a message indicated
  that daemonize is GPL-licensed. It's BSD-licensed.)
- Updated copyright information.

-----------------------------------------------------------------------------
Changes in version 1.5.2:

Two changes suggested and supplied by Yakov Lerner <iler.ml <at> gmail.com>:

- Added a "-l lockfile" option, permitting daemonize to use a lockfile to
  ensure that only one instance of the daemon runs at a time.

- Now makes additional checks to ensure that the program specified to
  daemonize is a regular file, not a directory.

-----------------------------------------------------------------------------
Changes in version 1.5.1:

- Now released under BSD-style license, not GNU Public License.

-----------------------------------------------------------------------------
Changes in version 1.5:

- daemonize can now optionally redirect the daemon's standard output and
  standard error to files

- configure now supports a local basename() emulator.

-----------------------------------------------------------------------------
Changes in version 1.4:

- Version of daemon(3) shipped with daemonize attempted to redirect
  stdin, stdout and stderr to /dev/null, by closing all file descriptors
  (using sysconf(_SC_OPEN_MAX) to determine the maximum file descriptor).
  This method wasn't working, and apparently the forked daemon process
  was reacquiring (or still had) valid stdin, stdout and stderr file
  descriptors. I made some changes to daemon.c to be more consistent with
  how the native FreeBSD and Linux versions behave.

  This bug may have contributed to a problem on AIX. According to
  Ron Avriel <ron /at/ cti2 /dot/ com>, when he telnets to an AIX 5.2
  system and uses daemonize create a daemon, he can't log out of the telnet
  session (i.e., the logout operation hangs) until he kills the daemon.
  That observation is consistent with the theory that the daemon process is
  still holding open the controlling terminal.

- daemonize now installs with 0755 permissions, instead of 0700.

-----------------------------------------------------------------------------
Changes in version 1.3:

- The "-c" option now actually works. Duh.

-----------------------------------------------------------------------------
Changes in version 1.2:

- The external names in getopt.c are now prefixed with "x_" (e.g.,
  x_getopt(), x_opterr, etc.). The function's name has been changed to
  avoid type conflicts with the native getopt() on the host operating
  system. (This is known to happen on Solaris 8, for instance.) Thanks
  to Michael Fuchs <mf /at/ mfox /dot/ de> for pointing out the problem
  on Solaris.

-----------------------------------------------------------------------------
Changes in version 1.1:

- Now uses local getopt.c unconditionally, instead of defaulting to
  whatever's on the system. This gets around a problem when installing on
  systems with GNU getopt (e.g., Linux). Unless POSIXLY_CORRECT is set in
  the environment, GNU getopt scans the entire command line, which means
  it'll complain about "unknown options" that are specified in the command
  being passed to daemonize. One solution is to have daemonize set
  POSIXLY_CORRECT right before calling getopt() (and then restore it before
  running the program-to-daemonize, in case it might affect that program's

daemonize.c
/*---------------------------------------------------------------------------*\
  NAME
        daemonize - run a command as a Unix daemon

  DESCRIPTION

        See accompanying man page for full details.

  LICENSE

        This source code is released under a BSD-style license. See the
        LICENSE file for details.

  Copyright (c) 2003-2010 Brian M. Clapper, bmc@clapper.org
\*---------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <pwd.h>
#include <stdarg.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/file.h>
#include "config.h"

/*---------------------------------------------------------------------------*\
                                 Constants
\*---------------------------------------------------------------------------*/

#define VERSION "1.7.2"

/*---------------------------------------------------------------------------*\
                                  Globals
\*---------------------------------------------------------------------------*/

static const char  *pid_file   = NULL;
static const char  *out_file   = NULL;
static const char  *err_file   = NULL;
static const char  *lock_file  = NULL;
static bool         be_verbose = FALSE;
static const char  *user       = NULL;
static char       **cmd        = NULL;
static const char  *cwd        = "/";
static int          null_fd    = -1;
static int          out_fd     = -1;
static int          err_fd     = -1;
static int          append     = 0;
static const char **env        = NULL;

/*---------------------------------------------------------------------------*\
                             Private Functions
\*---------------------------------------------------------------------------*/

/**
 * Akin to perl's die() function, this function prints the specified message
 * and exits the program with a non-zero error code.
 *
 * Parameters:
 *     format - a printf format string
 *     ...    - any additional arguments as necessary to satisfy the format
 */
static void die(const char *format, ...)
{
    va_list ap;

    va_start(ap, format);
    vfprintf(stderr, format, ap);
    va_end(ap);

    exit(1);
}

/**
 * Emits a message to standard output if the be_verbose flag is set.
 *
 * Parameters:
 *     format - a printf format string
 *     ...    - any additional arguments as necessary to satisfy the format
 */
static void verbose(const char *format, ...)
{
    va_list ap;

    if (be_verbose)
    {
        va_start(ap, format);
        vfprintf(stdout, format, ap);
        va_end(ap);
    }
}

/**
 * Emit the usage message and abort the program with a non-zero exit code.
 *
 * Parameters:
 *     prog - name of program, from argv[0]
 */
static void usage(char *prog)
{
    static const char *USAGE[] = 
    {
"Usage: %s [OPTIONS] path [arg] ...",
"",
"OPTIONS",
"",
"-a             Append to, instead of overwriting, output files. Ignored ",
"               unless -e and/or -o are specified.",
"-c <dir>       Set daemon's working directory to <dir>.",
"-e <stderr>    Send daemon's stderr to file <stderr>, instead of /dev/null.",
"-E var=value   Pass environment setting to daemon. May appear multiple times.",
"-o <stdout>    Send daemon's stdout to file <stdout>, instead of /dev/null.",
"-p <pidfile>   Save PID to <pidfile>.",
"-u <user>      Run daemon as user <user>. Requires invocation as root.",
"-l <lockfile>  Single-instance checking using lockfile <lockfile>.",
"-v             Issue verbose messages to stdout while daemonizing."
    };

    int i;

    prog = basename(prog);
    fprintf(stderr, "%s, version %s\n", prog, VERSION);
    for (i = 0; i < sizeof(USAGE) / sizeof(const char *); i++)
    {
        fprintf(stderr, USAGE[i], prog);
        fputc('\n', stderr);
    }

    exit(1);
}

/**
 * Add a string of the name "name=value" to the environment. Aborts on error.
 *
 * Parameters:
 *     opt    - option character, for errors
 *     envvar - string of the form name=value
 */
static void add_to_env(char opt, const char *envvar)
{
    char *eq = strchr(envvar, '=');
    if (eq == NULL)
        die("Argument to -%c (\"%s\") is not of the form name=value.\n",
            opt, envvar);

    /* Split the string into its name/value substrings. */

    int name_len = (int) (eq - envvar);
    int val_len  = (strlen(envvar) - name_len) - 1;

    char *name = (char *) malloc(name_len + 1);
    char *value = (char *) malloc(val_len + 1);
    (void) strncat(name, envvar, name_len);
    eq++;
    (void) strncat(value, eq, val_len);
    setenv(name, value, 1);
}

/**
 * Parse the command-line parameters, setting the various globals that are
 * affected by them.
 *
 * Parameters:
 *     argc - argument count, as passed to main()
 *     argv - argument vector, as passed to main()
 */
static void parse_params(int argc, char **argv)
{
    int  opt;
    int  argsLeft;

    opterr = 0;

    /*
      NOTE: x_getopt() is the old public domain getopt(). The source lives
      in "getopt.c". The function's name has been changed to avoid
      conflicts with the native getopt() on the host operating system. So
      far, I've seen two kinds of conflicts:

      1. GNU getopt() (e.g., on Linux systems) behaves differently from the
         old getopt(), unless POSIXLY_CORRECT is defined in the
         environment. Specifically, it insists on processing options even
         after the first non-option argument has been seen on the command
         line. Initially, I got around this problem by forcing the use of
         the included public domain getopt() function.

      2. The types used in the included public domain getopt() conflict with
         the types of the native getopt() on some operating systems (e.g.,
         Solaris 8).

      Using x_getopt() ensures that daemonize uses its own version, which
      always behaves consistently.
    */
    while ( (opt = x_getopt(argc, argv, "ac:u:p:vo:e:E:l:")) != -1)
    {
        switch (opt)
        {
            case 'a':
                append = 1;
                break;

            case 'c':
                cwd = x_optarg;
                break;

            case 'p':
                pid_file = x_optarg;
                break;

            case 'v':
                be_verbose = TRUE;
                break;

            case 'u':
                user = x_optarg;
                break;

            case 'o':
                out_file = x_optarg;
                break;

            case 'e':
                err_file = x_optarg;
                break;

            case 'l':
                lock_file = x_optarg;
                break;

            case 'E':
                add_to_env('E', x_optarg);
                break;

            default:
                fprintf(stderr, "Bad option: -%c\n", x_optopt);
                usage(argv[0]);
        }
    }

    argsLeft = argc - x_optind;
    if (argsLeft < 1)
        usage(argv[0]);

    cmd  = &argv[x_optind];
    return;
}

/**
 * Switch the process's effective and real user IDs to the ID associated with
 * the specified user name. Also switches to that user's group ID to the
 * primary group associated with the user.
 *
 * This function calls die() if anything fails.
 *
 * Parameters:
 *     user_name - name of user to which to switch
 *     uid       - current process user ID
 *     pid_file  - if non-NULL, specifies PID file path, ownership of which
 *                 should be changed to the user
 */
static void switch_user(const char *user_name, uid_t uid, const char *pid_file)
{
    struct  passwd *pw;

    if (uid != 0)
        die("Must be root to specify a different user.\n");

    if ( (pw = getpwnam(user_name)) == NULL )
        die("Can't find user \"%s\" in password file.\n", user_name);

    if (setgid(pw->pw_gid) != 0)
        die("Can't set gid to %d: %s\n", pw->pw_gid, strerror (errno));

    /*
      For systems supporting multiple group memberships, make sure ALL
      groups are added to the process, just in case someone depends on them.

      Patch by Ken Farnen <kenf@14Cubed.COM>, 18 February 2010. Modified
      to put the #ifdef in the config.h header, rather than in here.
    */
    if (initgroups(pw->pw_name, pw->pw_gid) == -1)
        die("Can't initialize secondary groups for \"%s\": %s\n",
            pw->pw_name, strerror (errno));

    if (pid_file != NULL)
    {
        verbose("Changing ownership of PID file to \"%s\" (%d)\n",
                user, pw->pw_uid);
        if (chown(pid_file, pw->pw_uid, pw->pw_gid) == -1)
        {
            die("Can't change owner of PID file \"%s\" to \"%s\" (%d): %s\n",
                pid_file, user, pw->pw_uid, strerror (errno));
        }
    }

    if (setegid(pw->pw_gid) != 0)
        die("Can't set egid to %d: %s\n", pw->pw_gid, strerror (errno));

    if (setuid(pw->pw_uid) != 0)
        die("Can't set uid to %d: %s\n", pw->pw_uid, strerror (errno));

    if (seteuid(pw->pw_uid) != 0)
        die("Can't set euid to %d: %s\n", pw->pw_uid, strerror (errno));

    /*
      Initialize environment to match new username.
      Patch by Ken Farnen <kenf@14Cubed.COM>, 18 February 2010.
    */
    setenv("USER", pw->pw_name,1);
    setenv("LOGNAME", pw->pw_name,1);
    setenv("HOME", pw->pw_dir,1);
}

/**
 * Open the specified output file. If the global "append" variable is set,
 * the file is opened in append mode; otherwise, it is overwritten.
 *
 * Parameters:
 *     path - path to the output file
 *
 * Returns:
 *     the file descriptor for the open file, or -1 on error.
 */
static int open_output_file(const char *path)
{
    int flags = O_CREAT | O_WRONLY;

    if (append)
    {
        verbose("Appending to %s\n", path);
        flags |= O_APPEND;
    }

    else
    {
        verbose("Overwriting %s\n", path);
        flags |= O_TRUNC;
    }

    return open(path, flags, 0666);
}

/**
 * Opens all output files.
 */
static void open_output_files()
{
    /* open files for stdout/stderr */

    if ((out_file != NULL) || (err_file != NULL))
    {
        if ((null_fd = open("/dev/null", O_WRONLY)) == -1)
            die("Can't open /dev/null: %s\n", strerror (errno));

        close(STDIN_FILENO);
        dup2(null_fd, STDIN_FILENO);

        if (out_file != NULL)
        {
            if ((out_fd = open_output_file(out_file)) == -1)
            {
                die("Can't open \"%s\" for stdout: %s\n", 
                    out_file, strerror(errno));
            }
        }

        else 
        {
            out_fd = null_fd;
        }

        if (err_file != NULL)
        {
            if ((out_file != NULL) && (strcmp(err_file, out_file) == 0))
                err_fd = out_fd;

            else if ((err_fd = open_output_file(err_file)) == -1)
            {
                die("Can't open \"%s\" for stderr: %s\n", 
                    err_file, strerror (errno));
            }
        }

        else
        {
            err_fd = null_fd;
        }
    }
}

/**
 * Redirects standard output and error to someplace safe.
 */
static int redirect_stdout_stderr()
{
    int rc = 0;

    /* Redirect stderr/stdout */

    if ((out_file != NULL) || (err_file != NULL))
    {
        close(STDIN_FILENO);
        close(STDOUT_FILENO);
        close(STDERR_FILENO);
        dup2(null_fd, STDIN_FILENO);
        dup2(out_fd, STDOUT_FILENO);
        dup2(err_fd, STDERR_FILENO);
        rc = 1;
    }

    return rc;
}

/*---------------------------------------------------------------------------*\
                               Main Program
\*---------------------------------------------------------------------------*/

int main(int argc, char **argv)
{
    uid_t  uid = getuid();
    int    noclose = 0;
    int    lockFD;
    struct stat st;

    if (geteuid() != uid)
        die("This executable is too dangerous to be setuid.\n");

    parse_params(argc, argv);

    if (cmd[0][0] != '/')
        die("The 'path' parameter must be an absolute path name.\n");

    /* Verify that the path to the command points to an existing file. */

    if (access(cmd[0], X_OK) == -1)
        die("File \"%s\" is not executable.\n", cmd[0]);

    /*
      Note: A directory will also pass the X_OK test, so test further with
      stat().
    */
    if (stat(cmd[0], &st))
        die("File \"%s\" does not exist.\n", cmd[0]);

    if(! S_ISREG(st.st_mode))
        die("File \"%s\" is not regular file.\n", cmd[0]);

    if (lock_file)
    {
        lockFD = open(lock_file, O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR);
        if (lockFD < 0)
            die("Can't create lock file \"%s\": %s\n",
                lock_file, strerror (errno));
        if (flock(lockFD, LOCK_EX | LOCK_NB) != 0)
            die("Can't lock the lock file \"%s\". "
                "Is another instance running?\n",
                lock_file);
    }

    if (pid_file != NULL)
    {
        int fd;

        verbose("Creating PID file \"%s\".\n", pid_file);
        fd = open(pid_file, O_CREAT | O_WRONLY,
                  S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
        if (fd < 0)
        {
            die ("Can't create PID file \"%s\": %s\n",
                 pid_file, strerror (errno));
        }

        close(fd);
    }

    if (user != NULL)
        switch_user(user, uid, pid_file);

    open_output_files();

    if (chdir (cwd) != 0)
    {
        die("Can't change working directory to \"%s\": %s\n",
            cwd, strerror (errno));
    }

    verbose("Daemonizing...");

    if (redirect_stdout_stderr())
        noclose = 1;

    if (daemon (1, noclose) != 0)
        die("Can't daemonize: %s\n", strerror (errno));

    if (chdir(cwd) != 0)
    {
        die("Can't change working directory to \"%s\": %s\n",
            cwd, strerror (errno));
    }

    if (pid_file != NULL)
    {
        FILE  *fPid = NULL;

        verbose("Writing process ID to \"%s\".\n", pid_file);
        if ( (fPid = fopen (pid_file, "w")) == NULL )
        {
            die("Can't open PID file \"%s\": %s\n",
                pid_file, strerror (errno));
        }

        fprintf(fPid, "%d\n", getpid());
        fclose(fPid);
    }

    /*
      Make sure we have a relatively sane environment
      Patch by Ken Farnen <kenf@14Cubed.COM>, 18 February 2010.
    */
    if (getenv("IFS") == NULL)
        setenv("IFS"," \t\n",1);

    if (getenv("PATH") == NULL)
        setenv("PATH","/usr/local/sbin:/sbin:/bin:/usr/sbin:/usr/bin", 1);

    execvp(cmd[0], cmd);

    die("Can't exec \"%s\": %s\n", cmd[0], strerror (errno));
}

/*  vim: set et sw=4 sts=4 : */

